### DISCUSSION TITLE
Question in 'bo_with_warped_gp.ipynb'

### DISCUSSION THREAD

**Discussion by HanyangHenry-Wang** [meta-pytorch/botorch#1802] [Q&A]
Hi,

I am trying to use warped GP to do BO and I found the function 'obj() ' in 'bo_with_warped_gp.ipynb' confuses me a bit. 

From my understanding, this function takes 'train_x_ei' as inputs and then calculates the function value 'train_obj_ei' after input warping. However, in the next step, the tutorial uses 'train_x_ei' and 'train_obj_ei' to train GP for normal BO (NEI). I do not really understand why we should use train_x_ei and train_obj_ei to build the GP because train_obj_ei is not the value of train_x_ei.

Could you please give me any hints? Thank you for your help.

**Comment by Balandat** [comment]
> because train_obj_ei is not the value of train_x_ei

I am not sure I understand; is your concern that `train_obj_ei` is not `obj(train_x_ei)` but a perturbed version of it (via `train_obj = exact_obj + NOISE_SE * torch.randn_like(exact_obj)`)? If so, this is just so that we can simulate a setting in which we only have access to noisy observations.

**Reply by HanyangHenry-Wang** [reply]
Thank you for your quick reply!
I am not worrying about the noise. What I do not understand is that the NEI part seems to build a GP model for (x,f(w(x))), where 'f' is the objective function (neg_hartmann6) and 'w' is warping function (k.icdf), but usually we will build surrogate model for (x,f(x)) in BO?

**Reply by Balandat** [reply]
So that's the whole point of the warping - apply a transformation `w` to the inputs and thus model `x -> f(w(x))`. The idea is that if `f` is hard to model with a GP (e.g. b/c the local variability of the function varies across the space), applying such an input transformation `w` could render it easier to model the composition `f \circ w`.

**Reply by HanyangHenry-Wang** [reply]
Thank you for your reply. Yes, I agree that if we use 'NEI+Input Warping', we need to transform the input space in the kernel. But if we use 'NEI' (which I assume to be normal BO), should we still warp the inputs?

My question is that why we still need to warp the inputs when we are using a normal BO?

**Reply by Balandat** [reply]
Hmm I see - yeah I guess looking a  bit closer at the tutorial, what we're doing is to just pick some random parameters of the Kumaraswamy warping function in the beginning, and use that to define a new objective function based on that. We then use that objective function as our "new f". So long as we use that to evaluate all methods we're good. The point of the tutorial is that by defining input warping on the model (which we only do for the `_warp` version as you can see from the arguments passed to `initialize_model` , we are able to learn those random parameter samples and get better performance as a result. We can probably be a bit more explicit about that in the tutorial comments. 

Does that make sense?

**Reply by HanyangHenry-Wang** [reply]
Got it! Yes, I think a bit more explanation in comments will be helpful for other users. Thank you for your help.
